### DISCUSSION TITLE
verbose `SIDE_MODULE=1` to get `including lib[...]`

### DISCUSSION THREAD

**Discussion by Keillion** [emscripten-core/emscripten#17216] [Q&A] [answered]
Thanks for reading.

<br>

It's a test cmake(emcmake) project. I try to find out what libs are needed by sideModule, following this [guide](https://emscripten.org/docs/compiling/Dynamic-Linking.html#system-libraries), but fail.

As `sideModule.cpp` writing, it includes `std string`. But I can't find any info in output.

I have tried use several ways to make it verbose:

* set(EM_BUILD_VERBOSE 1)
* set(CMAKE_VERBOSE_MAKEFILE 1)
* set(VERBOSE 1)
* set_target_properties(sideModule PROPERTIES COMPILE_FLAGS " -v -s SIDE_MODULE=1" LINK_FLAGS " -v -s SIDE_MODULE=1")
* emmake make VERBOSE=1

<br>

**Version of emscripten/emsdk:**
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.13 (531257621816c200bc7c3be53129494afd029aec)
clang version 15.0.0 (https://github.com/llvm/llvm-project 5c6ed60c517c47b25b6b25d8ac3666d0e746b0c3)

<br>

**Failing command line in full:**
https://github.com/Keillion/emsdk-cheatlist/tree/70b54c5f906df4cac5ac14adb23b8990c60552f8
```
$ path/to/emsdk/upstream/emscripten/tools/webidl_binder mainModule.idl mainModuleGlue
$ emcmake cmake .
$ emmake make VERBOSE=1
```

<br>

**Full link command and output with `-v` appended:**
I can't find any output has `including lib[...]`.
```log
...
...
(hide info about other targets, let's watch the sideModule)
...
...
make: make VERBOSE=1
/usr/bin/cmake -S/root/emsdk-cheatlist -B/root/emsdk-cheatlist --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /root/emsdk-cheatlist/CMakeFiles /root/emsdk-cheatlist/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/root/emsdk-cheatlist'
make -f CMakeFiles/sideModule.dir/build.make CMakeFiles/sideModule.dir/depend
make[2]: Entering directory '/root/emsdk-cheatlist'
cd /root/emsdk-cheatlist && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /root/emsdk-cheatlist /root/emsdk-cheatlist /root/emsdk-cheatlist /root/emsdk-cheatlist /root/emsdk-cheatlist/CMakeFiles/sideModule.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/root/emsdk-cheatlist'
make -f CMakeFiles/sideModule.dir/build.make CMakeFiles/sideModule.dir/build
make[2]: Entering directory '/root/emsdk-cheatlist'
make[2]: Nothing to be done for 'CMakeFiles/sideModule.dir/build'.
make[2]: Leaving directory '/root/emsdk-cheatlist'
[100%] Built target sideModule
make[1]: Leaving directory '/root/emsdk-cheatlist'
/usr/bin/cmake -E cmake_progress_start /root/emsdk-cheatlist/CMakeFiles 0
```

**Comment by sbc100** [comment, marked as answer]
I'm afraid it looks like that document is out of date.   These days that most efficient way to link in exactly that is needed into the main module is to build with `MAIN_MODULE=2` and pass the side module as an argument.    Are you able to do this?  (i.e. can you build your side module first and then pass it in when linking your main module?)

**Reply by sbc100** [reply]
PR to update the docs: https://github.com/emscripten-core/emscripten/pull/17272

**Reply by Keillion** [reply]
> pass it in when linking your main module

I guess you means this way.
```
emcc -sMAIN_MODULE=2 mainModule.cpp sideModule.wasm
```
Modify from [Load-time Dynamic Linking](https://github.com/emscripten-core/emscripten/blob/911c848cc26eccc55696ee1d8d6f17bb79720138/site/source/docs/compiling/Dynamic-Linking.rst#load-time-dynamic-linking)

---

In CMakeLists.txt, I try this:
```
set_target_properties(mainModule PROPERTIES COMPILE_FLAGS " -s MAIN_MODULE=2 <other_params>" LINK_FLAGS " -s MAIN_MODULE=2 <other_params> sideModule.wasm")
```
The compilation results are greatly reduced in size, and can run normally.

---

It seems `emmake make` can still process when sideModule.wasm not exists.
```
make: make
[ 33%] Built target mainModule
[ 50%] Linking CXX executable sideModule.wasm # this line not appears when sideModule.wasm already exists
[ 66%] Built target sideModule
[100%] Built target A
```
Not sure if there are any hidden dangers in doing so.
